bionic/task_state.py

```diff
@@ -317,3 +329,54 @@ def _check_accessors_for_version_problems(self):
 
         for accessor in accessors_needing_saving:
             accessor.update_provenance()
+
+    # NOTE: This can be optimized further.
+    def new_state_for_completion(self, new_task_states_by_key, to_be_computed=True):
+        """
+        Returns a copy of task state after keeping only the necessary data
+        required for completion. Along with keeping only necessary data for
+        completion, this also removes all the memoized results since they can
+        be expensive to serialize.
+
+        Mainly used to reduce IPC overhead when sending the state over to a
+        subprocess.
+
+        Parameters
+        ----------
+
+        new_task_states_by_key: Dict from key to new task states.
+            The cache for tracking new task states. Should be an empty dict when
+            first called.
+
+        to_be_computed: Boolean.
+            Whether the task state will be computed in the subprocess.
+        """
+
+        # All task keys should point to the same task state.
+        if self.task.keys[0] in new_task_states_by_key:
+            return new_task_states_by_key[self.task.keys[0]]
+
+        # Let's make a copy of the task state.
+        # Note that this is not a deep copy so don't mutate so be careful when
+        # mutating state variables.
+        task_state = copy.copy(self)
+
+        # Clear up memoized cache to avoid sending it through IPC.
+        task_state._results_by_name = None
+
+        if to_be_computed:
+            new_dep_states = []
+            for dep_state in task_state.dep_states:
+                # We will need to compute the dependency in subprocess if it's not persisted.
```

Comment:
phrasing: "in subprocess" is not grammatical. Maybe "in the other subprocess"? Also, might be clearer to say "recompute" instead of "compute"?

Likewise a few lines down.